lib/auth/join_azure.go

```diff
@@ -0,0 +1,398 @@
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+	if intermediate != nil {
+		p7.Certificates = append(p7.Certificates, intermediate)
+	}
+
+	pool := x509.NewCertPool()
+	for _, cert := range certs {
+		pool.AddCert(cert)
+	}
+
+	if err := p7.VerifyWithChain(pool); err != nil {
+		return "", trace.Wrap(err)
+	}
+
+	return ad.ID, nil
+}
+
+// verifyVMIdentity verifies that the provided access token came from the
+// correct Azure VM.
+func verifyVMIdentity(ctx context.Context, cfg *azureRegisterConfig, accessToken, vmID string, requestStart time.Time) (*azure.VirtualMachine, error) {
+	tokenClaims, err := cfg.verify(ctx, accessToken)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	expectedIssuer, err := url.JoinPath("https://sts.windows.net", tokenClaims.TenantID, "/")
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	// v2 tokens have the version appended to the issuer.
+	if tokenClaims.Version == "2.0" {
+		expectedIssuer, err = url.JoinPath(expectedIssuer, "2.0")
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+	}
+
+	expectedClaims := jwt.Expected{
+		Issuer:   expectedIssuer,
+		Audience: jwt.Audience{azureAccessTokenAudience},
+		Time:     requestStart,
+	}
+
+	if err := tokenClaims.Validate(expectedClaims); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	rsID, err := arm.ParseResourceID(tokenClaims.ResourceID)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	vmClient := cfg.vmClient
+	if vmClient == nil {
+		tokenCredential := azure.NewStaticCredential(azcore.AccessToken{
+			Token:     accessToken,
+			ExpiresOn: tokenClaims.Expiry.Time(),
+		})
+		var err error
+		vmClient, err = azure.NewVirtualMachinesClient(rsID.SubscriptionID, tokenCredential, nil)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+	}
+
+	resourceID, err := arm.ParseResourceID(tokenClaims.ResourceID)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	var vm *azure.VirtualMachine
+
+	// If the token is from the system-assigned managed identity, the resource ID
+	// is for the VM itself and we can use it to look up the VM.
+	if slices.Contains(resourceID.ResourceType.Types, "virtualMachines") {
+		vm, err = vmClient.Get(ctx, tokenClaims.ResourceID)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		if vm.VMID != vmID {
+			return nil, trace.AccessDenied("vm ID does not match")
+		}
+
+		// If the token is from a user-assigned managed identity, the resource ID is
+		// for the identity and we need to look the VM up by VM ID.
+	} else {
+		vm, err = vmClient.GetByVMID(ctx, resourceID.ResourceGroupName, vmID)
+		if err != nil {
+			if trace.IsNotFound(err) {
+				return nil, trace.AccessDenied("no VM found with matching VM ID")
+			}
+			return nil, trace.Wrap(err)
+		}
+	}
+
+	return vm, nil
+}
+
+func checkAzureAllowRules(vm *azure.VirtualMachine, allowRules []*types.ProvisionTokenSpecV2Azure_Rule) error {
+	for _, rule := range allowRules {
+		if len(rule.Subscription) > 0 {
```

Comment:
If we always require `Subscription` to be configured (as CheckAndSetDefaults for ProvisionTokenSpecV2Azure states) , we could conform with that here by removing this if statement ?